src/shogun/io/ARFFFile.cpp

```diff
@@ -0,0 +1,357 @@
+					auto pos =
+					    std::find(encoding.begin(), encoding.end(), trimmed_el);
+					if (pos == encoding.end())
+						SG_SERROR(
+						    "Unexpected value \"%s\" on line %d\n",
+						    trimmed_el.c_str(), m_line_number);
+					float64_t idx = std::distance(encoding.begin(), pos);
+					shogun::get<std::vector<float64_t>>(m_data_vectors[i])
+					    .push_back(idx);
+					nominal_pos = std::next(nominal_pos);
+				}
+				break;
+				case (Attribute::Date):
+				{
+					date::sys_seconds t;
+					std::istringstream ss(elems[i]);
+					if (date_pos == m_date_formats.end())
+						SG_SERROR(
+						    "Unexpected date value \"%s\" on line %d.\n",
+						    elems[i].c_str(), m_line_number);
+					ss >> date::parse(*date_pos, t);
+					if (bool(ss))
+					{
+						auto value_timestamp = t.time_since_epoch().count();
+						shogun::get<std::vector<float64_t>>(m_data_vectors[i])
+						    .push_back(value_timestamp);
+					}
+					else
+						SG_SERROR(
+						    "Error parsing date \"%s\" with date format \"%s\" "
+						    "on line %d.\n",
+						    elems[i].c_str(), (*date_pos).c_str(),
+						    m_line_number)
+					++date_pos;
+				}
+				break;
+				case (Attribute::String):
+					shogun::get<std::vector<std::string>>(m_data_vectors[i])
+					    .emplace_back(elems[i]);
+				}
+			}
+			if (i != m_attributes.size())
+				SG_SERROR(
+				    "Unexpected number of values on line %d, expected %d "
+				    "values, "
+				    "but found %d.\n",
+				    m_line_number, m_attributes.size(), i)
+			++m_row_count;
+		}
+	};
+	auto check_data = [this]() {
+		// check X values
+		SG_SDEBUG(
+		    "size: %d, cols: %d, rows: %d", m_data_vectors.size(),
+		    m_data_vectors.size() / m_row_count, m_row_count)
+		if (!m_data_vectors.empty())
+		{
+			auto feature_count = m_data_vectors.size();
+			index_t row_count =
+			    shogun::visit(VectorSizeVisitor{}, m_data_vectors[0]);
+			for (int i = 1; i < feature_count; ++i)
+			{
+				REQUIRE(
+				    shogun::visit(VectorSizeVisitor{}, m_data_vectors[i]) ==
+				        row_count,
+				    "All columns must have the same number of features!\n")
+			}
+		}
+		else
+			return false;
+		return true;
+	};
+	process_chunk(read_data, check_data, true);
+}
+
+std::shared_ptr<CCombinedFeatures> ARFFDeserializer::get_features() const
+{
+	auto result = std::make_shared<CCombinedFeatures>();
+	index_t row_count = shogun::visit(VectorSizeVisitor{}, m_data_vectors[0]);
+	for (int i = 0; i < m_data_vectors.size(); ++i)
+	{
+		Attribute att = m_attributes[i];
+		auto vec = m_data_vectors[i];
+		switch (att)
+		{
+		case Attribute::Numeric:
+		case Attribute::Integer:
+		case Attribute::Real:
+		case Attribute::Date:
+		case Attribute::Nominal:
+		{
+			auto casted_vec = shogun::get<std::vector<float64_t>>(vec);
+			SGMatrix<float64_t> mat(1, row_count);
+			memcpy(
+			    mat.matrix, casted_vec.data(),
+			    casted_vec.size() * sizeof(float64_t));
+			auto* feat = new CDenseFeatures<float64_t>(mat);
```

Comment:
do we want everything to be float64?

Reply:
mmmh, I can add `int64_t` too when a value is specifically `integer`

Reply:
my question is more: what if users want float32/int32 or whatever .... see loading csv files

Reply:
ah sure, well the user would have to specify int and float type imo, i.e. float32/int32 or float32/int64

Reply:
at least there should be an API for that. And then some kind of gatekeeper function offers that to users ?

Reply:
as discussed in IRC, I think it might make sense to template the primitive type, and then give a API that takes an enum (the usual `PT_FLOAT64` etc) to select how to cast the numbers

Reply:
yes, I was trying to find that in csv but couldn't find it? I thought they return type depended on the getter used? i.e. `get_vector(T*& vector, int32_t& len);`

Reply:
I meant the factory, e.g. `CFeatures* features(CFile* file, machine_int_t primitive_type = PT_FLOAT64)`